derive/src/lib.rs

```diff
@@ -28,69 +28,89 @@ use proc_macro::TokenStream;
 use quote::quote;
 use syn::parse_macro_input;
 use syn::Data;
+use syn::DeriveInput;
+use syn::GenericParam;
 use syn::ItemFn;
 
 #[proc_macro_derive(AlignedBorrow)]
 pub fn aligned_borrow_derive(input: TokenStream) -> TokenStream {
-    let ast: syn::DeriveInput = syn::parse(input).unwrap();
-
-    // Get struct name from ast
+    let ast = parse_macro_input!(input as DeriveInput);
     let name = &ast.ident;
-    let methods = quote! {
-        impl<T> Borrow<#name<T>> for [T] {
-            fn borrow(&self) -> &#name<T> {
-                debug_assert_eq!(self.len(), size_of::<#name<u8>>());
-                let (prefix, shorts, _suffix) = unsafe { self.align_to::<#name<T>>() };
-                debug_assert!(prefix.is_empty(), "Alignment should match");
-                debug_assert_eq!(shorts.len(), 1);
-                &shorts[0]
-            }
-        }
 
-        impl<T> BorrowMut<#name<T>> for [T] {
-            fn borrow_mut(&mut self) -> &mut #name<T> {
-                debug_assert_eq!(self.len(), size_of::<#name<u8>>());
-                let (prefix, shorts, _suffix) = unsafe { self.align_to_mut::<#name<T>>() };
-                debug_assert!(prefix.is_empty(), "Alignment should match");
-                debug_assert_eq!(shorts.len(), 1);
-                &mut shorts[0]
+    // Separate type generics and const generics
+    let type_generic = ast
+        .generics
+        .params
+        .iter()
+        .find_map(|param| {
+            if let GenericParam::Type(type_param) = param {
+                Some(&type_param.ident)
+            } else {
+                None
+            }
+        })
+        .expect("Expected at least one type generic");
+
```

Comment:
Could you also add a panic! that would occur if a non-const generic is used other than T? that would be really good for our user experience which might not be clear right now.

Reply:
Please see #368. Infact, I have modified the code to trigger a panic if the first generic is not a type, and for any subsequent generics, if they are not consts.